### PR TITLE
allow enabling GC instead of bump-ptr allocations via command line switch -GC

### PR DESCRIPTION
**Pull request by rainers**
Alternative to https://github.com/dlang/dmd/pull/9048 reusing most of @kinke changes to rmem.d.

Should best be compiled with a recent druntime to reduce memory overhead for GC allocations: https://github.com/dlang/druntime/pull/2384

### CODE REVIEW

**Comment by rainers** [comment]
Oops, the makefiles build via a library so they are hit by https://github.com/dlang/dmd/pull/9132.

File: src/dmd/mars.d [hunk 1]
```diff
@@ -2374,3 +2381,34 @@ Modules createModules(ref Strings files, ref Strings libmodules)
...
     return modules;
 }
+
+private
+{
+    version (GNU)
+    {
+        // does not support pragma(crt_constructor) yet
+        import gcc.attribute;
+        @attribute("section", ".ctors") auto pini = &initGC;
```
Comment (jacob-carlborg) [inline src/dmd/mars.d]: This is ELF specific.

File: src/dmd/mars.d [hunk 2]
```diff
@@ -2374,3 +2381,34 @@ Modules createModules(ref Strings files, ref Strings libmodules)
...
+        extern(C) pragma(crt_constructor) void crt_initGC()
+        {
+            initGC(0, null);
+        }
+    }
+    else version (LDC)
+    {
+        // does not allow arguments on pragma(crt_constructor)
+        import ldc.attributes : section;
+        @section(".ctors") auto pini = &initGC;
```
Comment (jacob-carlborg) [inline src/dmd/mars.d]: Same here, ELF specific.

**Comment by kinke** [comment]
I'm not convinced this is much of an improvement. - I also don't like the `-GC` switch; I tried to avoid any mentioning of GC, as people probably associate it with *their* code.

**Comment by nordlow** [comment]
Are there any benchmarks illustrating reductions in maximum memory usage when compiling typical projects?

**Comment by rainers** [comment]
> Are there any benchmarks illustrating reductions in maximum memory usage when compiling typical projects?

Latest results with LDC are here: https://github.com/ldc-developers/ldc/pull/2916#issuecomment-443433594

**Comment by rainers** [comment]
> I've now tried a hacky access to gc.config.config.disable

With the introduction of the ProtoGC the GC configuration is initialized too late, so that also doesn't work.

**Pull request closed without merging**